compiler/src/iree/compiler/GlobalOptimization/GlobalLoopInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/GlobalOptimization/PassDetail.h"
+#include "iree/compiler/GlobalOptimization/Passes.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/PatternMatch.h"
+
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "global-loop-invariant-code-motion"
+#define KD_DBGS() (llvm::dbgs() << '[' << DEBUG_TYPE << "] ")
+
+namespace mlir::iree_compiler::GlobalOptimization {
+
+// Check if the op is a leaf op we always want to hoist.
+static bool isHoistableLeafOp(Operation *op) {
+  // Currently it's limited to a small set of ops related to constant pack op.
+  return isa<tensor::PackOp>(op);
+}
+
+// Check if the op is hoistable (but we might not want to hoist it alone).
+static bool isHoistableOp(Operation *op) {
+  // Currently it's limited to a small set of ops related to constant pack op.
+  return op->hasTrait<OpTrait::ConstantLike>() || isa<tensor::EmptyOp>(op) ||
+         isHoistableLeafOp(op);
+}
+
+// Check if the op and its producers are loop invariants and hoistable. Results
+// are cached in hoistableOpMap to avoid repeated traversals.
+static bool
+checkHoistableTree(LoopLikeOpInterface loopOp, Operation *op,
+                   llvm::SmallDenseMap<Operation *, bool> &hoistableOpMap) {
+  // First check if the op has been analyzed.
+  if (hoistableOpMap.contains(op))
+    return hoistableOpMap[op];
+
+  // Check if the nested ops and the op itself are hoistable. And check if the
+  // producers of all operands are hoistable.
+  auto walkFn = [&](Operation *walkOp) {
+    // Check if the op is hoistable.
+    if (!isHoistableOp(walkOp))
+      return WalkResult::interrupt();
+
+    WalkResult result = WalkResult::advance();
+    // Check if the producers of operands are hoistable.
+    for (OpOperand &operand : walkOp->getOpOperands()) {
+      Value value = operand.get();
+      // Ignore values defined in a nested region or outside the loop.
+      if (walkOp->isAncestor(value.getParentRegion()->getParentOp()) ||
+          loopOp.isDefinedOutsideOfLoop(value)) {
+        continue;
+      }
+
+      Operation *producer = value.getDefiningOp();
+      // If the value is not an operation, we don't hoist it.
+      if (!producer || !checkHoistableTree(loopOp, producer, hoistableOpMap)) {
+        result = WalkResult::interrupt();
+        break;
+      }
+    }
+    return result;
```

Comment:
I think we also have to check that the operation we want to hoist is immediately nested in the loop. I'm not sure we can hoist operations out of an nested if statement, for example.

Reply:
When adding the hoistable ops at line 111, we only add immediately nested ops in the loop. I added a test to make sure we don't hoist ops from deeper than that.

Reply:
It would be good to add a comment about this there

Reply:
Added.